includes/classes/class-plugin-manager.php

```diff
@@ -0,0 +1,409 @@
+
+		$compatible_php = is_php_version_compatible( $requires_php );
+		$compatible_wp  = is_wp_version_compatible( $requires_wp );
+		$tested_wp      = ( empty( $plugin['tested'] ) || version_compare( get_bloginfo( 'version' ), $plugin['tested'], '<=' ) );
+		$action_links   = array();
+
+		if ( current_user_can( 'install_plugins' ) || current_user_can( 'update_plugins' ) ) {
+			$status = install_plugin_install_status( $plugin );
+
+			switch ( $status['status'] ) {
+				case 'install':
+					if ( $status['url'] ) {
+						if ( $compatible_php && $compatible_wp ) {
+							$action_links[] = sprintf(
+								'<a class="install-now button" data-slug="%s" href="%s" aria-label="%s" data-name="%s">%s</a>',
+								esc_attr( $plugin['slug'] ),
+								esc_url( $status['url'] ),
+								/* translators: %s: Plugin name and version. */
+								esc_attr( sprintf( _x( 'Install %s now', 'plugin', 'default' ), $name ) ),
+								esc_attr( $name ),
+								__( 'Install Now', 'default' )
+							);
+						} else {
+							$action_links[] = sprintf(
+								'<button type="button" class="button button-disabled" disabled="disabled">%s</button>',
+								_x( 'Cannot Install', 'plugin', 'default' )
+							);
+						}
+					}
+					break;
+
+				case 'update_available':
+					if ( $status['url'] ) {
+						if ( $compatible_php && $compatible_wp ) {
+							$action_links[] = sprintf(
+								'<a class="button aria-button-if-js" data-plugin="%s" data-slug="%s" href="%s" aria-label="%s" data-name="%s">%s</a>',
+								esc_attr( $status['file'] ),
+								esc_attr( $plugin['slug'] ),
+								esc_url( $status['url'] ),
+								/* translators: %s: Plugin name and version. */
+								esc_attr( sprintf( _x( 'Update %s now', 'plugin', 'default' ), $name ) ),
+								esc_attr( $name ),
+								__( 'Update Now', 'default' )
+							);
+						} else {
+							$action_links[] = sprintf(
+								'<button type="button" class="button button-disabled" disabled="disabled">%s</button>',
+								_x( 'Cannot Update', 'plugin', 'default' )
+							);
+						}
+					}
+					break;
+
+				case 'latest_installed':
+				case 'newer_installed':
+					if ( is_plugin_active( $status['file'] ) ) {
+						$action_links[] = sprintf(
+							'<button type="button" class="button button-disabled" disabled="disabled">%s</button>',
+							_x( 'Active', 'plugin', 'default' )
+						);
+						if ( current_user_can( 'deactivate_plugin', $status['file'] ) ) {
+							global $page;
+							$s       = isset( $_REQUEST['s'] ) ? $_REQUEST['s'] : ''; // phpcs:ignore
+							$context = $status['status'];
+
+							$action_links[] = sprintf(
+								'<a href="%s" id="deactivate-%s" aria-label="%s" style="color:red;text-decoration: underline;">%s</a>',
+								wp_nonce_url( 'plugins.php?wpp=1&action=deactivate&amp;plugin=' . rawurlencode( $status['file'] ) . '&amp;plugin_status=' . $context . '&amp;paged=' . $page . '&amp;s=' . $s, 'deactivate-plugin_' . $status['file'] ),
```

Comment:
We should rather use the more precise approach to create such a URL, similar to how the activate URL is assembled below:
```suggestion

							$deactivate_url = add_query_arg(
								array(
									'_wpnonce' => wp_create_nonce( 'deactivate-plugin_' . $status['file'] ),
									'action'   => 'deactivate',
									'plugin'   => $status['file'],
									'wpp'      => 1,
								),
								network_admin_url( 'plugins.php' )
							);

							$action_links[] = sprintf(
								'<a href="%s" id="deactivate-%s" aria-label="%s" style="color:red;text-decoration: underline;">%s</a>',
								esc_url( $deactivate_url ),
```

Additionally, see my related comment on the PHP handler for this though, on updating the request parameters.

Reply:
@felixarntz this has been addressed in the following commits:

- https://github.com/WordPress/performance/pull/864/commits/bb58fc5982cebc657dc234313a52e2aa55ed51e7
- https://github.com/WordPress/performance/pull/864/commits/ddaf667b43e9519612466ed7a9269c55fb1dcd1c
- https://github.com/WordPress/performance/pull/864/commits/2583f3d0581bc348e1d8f0c6e3c4ff2dc511b2d9